File: 1335-maximum-candies-allocated-to-k-children/1335-maximum-candies-allocated-to-k-children.py
class Solution:
    def maximumCandies(self, candies: List[int], k: int) -> int:
        l, r = 1, sum(candies)

        def work(mid):
            if mid == 0:
                return False
            res = 0  # if candies[i] can seperate by mid then res+1
            for c in candies:
                res += c // mid
            return res >= k

        while l <= r:
            mid = (l + r) // 2
            if work(mid):
                l = mid + 1
            else:
                r = mid - 1
        return r if r>=0 else 0

File: 1335-maximum-candies-allocated-to-k-children/1335-maximum-candies-allocated-to-k-children.cs
public class Solution {
    
    private bool Work(List<int> candies2,long c, long k){
        long res = 0;
        foreach(var i in candies2)
        {
            res += (i/c);
        }
        return true? res>=k :false;
    }
    
    public int MaximumCandies(int[] candies, long k) {
        List<int> candies2 = new List<int>(candies);
        
        long l = 1; 
        long r = candies2.Max();
        
        while(l<=r)
        {
            long mid = (l+r) / 2;
            if(Work(candies2,mid,k))
                l = mid+1;
            else
                r = mid-1;
        }
        return (int)r;
    }
}